kostyor/upgrades/engine.py

```diff
@@ -152,17 +152,21 @@ def start(self):
         # keystone, then with nova, and so on. See iteration details in
         # get_controllers() docstring.
         for host in get_controllers(hosts):
+            self.driver.pre_host_upgrade_hook(host)
             for service in iterservices(host):
+                self.driver.pre_service_upgrade_hook(service)
```

Comment:
The point of these hooks is that there _may_ be things that a driver does before a service or host is upgraded. We don't know, since it'd be up to the upgrade driver. This just gives the upgrade driver information about what is going on. 

I don't think putting everything inside of `start_upgrade` - including pre/post is a good practice.